### DISCUSSION TITLE
External Module - how to ceate db tables and install when adding external module to Oqtane

### DISCUSSION THREAD

**Discussion by markagama** [oqtane/oqtane.framework#1744] [Q&A]
Hi, probably another FAQ but I can't find the answer so I'll ask here.

I'm working on an external module created using Oqtane.

I would appreciate knowing if there is an example external module that demonstrates how an external module should utilise, manage, install database tables to the Oqtane database.

thx in advance

**Comment by markagama** [comment]
I found the great example (for an earlier version of Oqtane) by Michael Washington called Oqtane.Survey and have compared it with the external module code created automatically by Oqtane.

Oqtane.Survey loads the module by calling public SurveyManager(ISurveyRepository SurveyRepository, ISqlRepository sql)
Oqtane.Framework 2.2 external module calls public Module1Manager(IModule1Repository Module1Repository, ITenantManager tenantManager, IHttpContextAccessor accessor)

Now in the Oqtane.Server ModuleDefinitionController.cs I can see a db uninstall 
                                    _sql.ExecuteScript(tenant, moduletype.Assembly, Utilities.GetTypeName(moduledefinition.ModuleDefinitionName) + ".Uninstall.sql");

but I cannot find an equivalent install.sql line?

And I don't see where the _sql is being passed to the external module from Oqtane.Framework

Should I assume that the install.sql gets added to the module nuget package and this is used to install the db after the module is added to the production Oqtane.Framework? I see this in the Oqtane.Framework DatabaseManager

I'm still not seeing where the Oqtane.Framework _sql is being passed to the external module from Oqtane.Framework

Pointers would be appreciated.

**Reply by leigh-pointer** [reply]
Oqtane uses entity framework.   look in your module folder Server\Migrations\ 

Oqtane Framework no longer uses scripts due to multiple dB support.

**Reply by markagama** [reply]
thank you @leigh-pointer

this leads me to the question of a module adding tables to the Oqtane.Framework db when the module is installed?

I'll have a read and see what I find.

If you know of an example external module that uses the Oqtane.Framework v2.2+ db please let me know.

**Reply by leigh-pointer** [reply]
@markagama the external module template will generate your code plus the code to create / modify your data table.
login as host navigate to the admin dashboard and select "Module Management" here you can create a new module based on the default template.

Open the module with you dev env app and navigate to  Server\Migrations\ where you will find the code that creates the data table.

**Reply by sbwalker** [reply]
The Blog module https://github.com/oqtane/oqtane.blogs is also a good reference of how to use the EF Core migrations functionality to create tables.

**Reply by markagama** [reply]
@sbwalker thank you, I have this module and will look at this capability later this week.

**Reply by curelom** [reply]
How do we kick off the migrations?  I've added the InitializeModule class and entity builders to my existing module, but it doesn't create the tables.
